File: K Sized Subarray Maximum.cpp
class Solution {
  public:
    vector<int> max_of_subarrays(int k, vector<int> &arr) {
        deque<int>dq;
        vector<int>ans;
        int i=0,j=0,n=arr.size();
        while(j < n){
            while(dq.size() > 0 and arr[dq.front()] < arr[j]){
                dq.pop_front();
            }
            dq.push_front(j);
            if(j-i+1 < k){
                j++;
            }
            else{
                ans.push_back(arr[dq.back()]);
                if(i == dq.back()){
                    dq.pop_back();
                }
                i++;j++;
            }
        }
        
        return ans;
    }
};

File: K Sized Subarray Maximum.py
from collections import deque
class Solution:
    def maxOfSubarrays(self, arr, k):
        
        dq = deque()
        res = []
        
        for i, v in enumerate(arr):
            
            # drop items out of current window
            while dq and dq[0] <= i - k:
                dq.popleft()
            
            # drop items which violate constraint
            # monotonic decreasing deque
            while dq and arr[dq[-1]] < v:
                dq.pop()
                
            dq.append(i)
            
            if i >= k - 1:
                res.append(arr[dq[0]])
        
        return res
